BCprop - Add blade array switching

NoSloppy opened the pull request:
Should probably limit this to only if there's more than 1 blade config array?

- guard the array switching code, done.

NoSloppy commented:
> Should probably limit this to only if there's more than 1 blade config array?
> 
> * guard the array switching code, done.

Nope. Reverted.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -2042,6 +2064,154 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+  size_t prev_real_best_config = NELEM(blades);
+
+#ifdef BLADE_ID_SCAN_MILLIS
+  // Must be called from loop()
```
"must be called from loop" is the same comment as PollScanId() in prop base, but presumably this is not the same function, or you wouldn't have overridden it. Please update the comment to explain what this function does differently from the base function, otherwise the reader (me) will have to compare manually.

NoSloppy replied:
"must be called from loop"  was from the initial copy . removed.
Comments added to hopefully explain the functions.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -2042,6 +2064,154 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+  // Manual blade array selection
+  // even with Real TIme Blade ID active
+  bool manual_blade_array_active = false;
+  size_t last_real_best_config = NELEM(blades);
```
Each of these variables need a comment to explain what they do and why.

NELEM(blades) is probably not a valid value for these variables, I assum that is on purpose?

NoSloppy replied:
changed to SIZE_MAX

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -2042,6 +2064,154 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+  }
+
+  // Manual Blade Array Selection version of FindBladeAgain()
+  void FakeFindBladeAgain() {
```
Seems like it would be easier to just spoof the blade ID and then call the real FindBladeAgain()

NoSloppy replied:
Pretty sure I tried simple.
Activation and deactivation of blades i think were the issue. Possibly something else but I needed to duplicate some.
Sorry, a little rusty on this one at first glance.
I'l re-read the thread when this was going on to refresh.

NoSloppy replied:
> I'm going to see if I can come up with a simpler way.

That's great

NoSloppy commented:
I snuck in a couple of small edits to the instructions comments as opposed to a dedicated PR.

profezzorn commented:
This PR now has conflicts, also, have you considered this method:
https://crucible.hubbe.net/t/switching-between-many-more-than-2-preset-lists/6065/27
?

NoSloppy commented:
> have you considered this method: 

Playing with it now.
I assume you meant SaberBase -> PropBase, 
and `next_id_ = blades[n].id;` should be 'ohm'.

However, while this is much shorter/simpler, it doesn't seem to play nice with BladeID also being active (at first tests).
Regardless of what's manually set, it's always overridden by the reading.

That is the intent here: to have manual NextBladeArray, manual TriggerBladeID again, and automatic Blade ID active all at the same time.
That's what the code I submitted does well.
It was a lot of work to get it to do that.
I very well may have a bunch of superfluous stuff in the code, which can be optimized if possible, but it seems like the suggestion from the Crucible thread is too simplified to do all this?

Not sure you knew the full expectations of the code.

profezzorn commented:
Yes, I meant PropBase::id()

Did you make it virtual? (because the symptoms you describe is what would happen if it wasn't.)

> Not sure you knew the full expectations of the code.

No, I don't. The code duplication makes it really really difficult to tease out what has changed and what is merely duplicated, so I haven't done that.

NoSloppy commented:
> The code duplication makes it really really difficult to tease out what has changed and what is merely duplicated...

Ok so I refactored significantly, cleaning up where I could. By splitting PropBase::FindBlade() up into helper functions, I can just call them in the way needed, making it easier to read hopefully. 
I tried to incorporate your overriding id() suggestion, but it still requires the flags and tracking I had in place to allow for a spoofed Blade ID for manual switching while Blade ID is active.

I also unguarded the blade_detected_ bool (required BLADE_DETECT_PIN defined) , and removed additional changes not specifically relevant to this update. Those can be a small touchup PR to follow instead.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -1823,7 +1845,7 @@ void DoSavedTwist() {
     if (battle_mode_) {
       if (!hybrid_font.PlayPolyphonic(&SFX_bmbegin)) {
-        hybrid_font.DoEffect(EFFECT_FORCE, 0);
+        hybrid_font.PlayPolyphonic(&SFX_force);
```
I don't know why this shows up here as a diff and no conflict.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -2007,6 +2029,159 @@ void DoSavedTwist() {
+#ifdef BLADE_DETECT_PIN
+    if (!blade_detected_) {
+      PVLOG_STATUS << "NO ";
+      ret += NO_BLADE;
```
With this total override though, the base id() wouldn't get used, so I made sure to include all it does after the custom stuff here (without plain calling PropBase::id() because that would undermine the purpose here)

profezzorn replied:
But you're calling PropBase::id() on line 2082?

NoSloppy replied:
Oh I think I meant to put that before the early return of use_next_id is true. 
Getting sloppy. Also not at my computer at the moment.

NoSloppy replied:
If you're going to be doing something simpler, this might all change so not sure how much I should bother tweaking here.

NoSloppy commented on props/saber_BC_buttons.h:
```diff
@@ -2007,6 +2029,159 @@ void DoSavedTwist() {
+
+  // Play array.wavs if they exist and we manually switched.
+  // Fallback to playing font.wav.
+  void HandlePresetSound() {
```
renaming

NoSloppy commented:
Ok. Refactored again based on your suggestions. All seems to be working correctly in testing.
All array detection and switching works right.
Manually triggering a BLADE ID scan works right.
Blade in/out choosing correct config works right.

This was the desired behavior as far as sounds goes
- When manually switching arrays, array.wav files should be used if they exist, otherwise play bladein/out sounds. Fallback to font.wav
- When a REAL blade inserted or removed, bladein/out sounds should play. Fallback to font.wav

The only time the sound doesn't "work" is when you're already manually switched to an array which a real blade insertion/removal would also trigger (like being on array 2, and inserting a blade that triggers array 2.)
No sound is played in this scenario (I can live with that if need be?)

Otherwise, tested with/without the sounds, all fallback and work as expected otherwise.
Thanks.

profezzorn commented on props/prop_base.h:
```diff
@@ -668,6 +672,27 @@ class PropBase : CommandParser, Looper, protected SaberBase, public ModeInterfac
 #endif
   }
 
+  void FindBladeAgain() {
```
Looks like this was moved?
(Can we undo that to make the diff smaller?)

NoSloppy replied:
Yes I just moved it to group it with the other relative functions. I can move it back and add that last piece of puzzle tomorrow. Thanks.

NoSloppy replied:
moved back

profezzorn commented:
> The only time the sound doesn't "work" is when you're already manually switched to an array which a real blade insertion/removal would also trigger (like being on array 2, and inserting a blade that triggers array 2.) No sound is played in this scenario (I can live with that if need be?)

If you want to fix this, you would need to check in id() if returning real_id is going to result in a change or not. If it wouldn't you would set a bool variable, and then play a sound in loop() if that variable is set. (I don't think it's a good idea to to DoEffect() directly from id() since it's sometimes called from odd places.)

It would look something like this: 
```
  if (current_config - blade == real_best_config) play_no_change_sound_ = true;
```
(this would go right before "return real_id;")

Oh, one more thing; you would want to reset play_no_change_sound_  to false in PlayArraySound().

profezzorn commented on props/prop_base.h:
```diff
@@ -514,9 +514,9 @@ class PropBase : CommandParser, Looper, protected SaberBase, public ModeInterfac
   }
 
-#ifdef BLADE_DETECT_PIN
+// #ifdef BLADE_DETECT_PIN
```
Why comment out the #ifdef ?

NoSloppy replied:
I am finishing up the edits to my prop - needed blade_detected_ in there, but doing without it instead.
The aboves above will revert in a moment.

NoSloppy replied:
done.

NoSloppy commented on props/prop_base.h:
```diff
@@ -74,7 +74,7 @@ class PropBase : CommandParser, Looper, protected SaberBase, public ModeInterfac
 #ifdef MENU_SPEC_TEMPLATE
     MKSPEC<MENU_SPEC_TEMPLATE>::SoundLibrary::init();
-#endif    
+#endif
```
Sorry all these whitespace corrections showed up. I just overpasted the whole file.

NoSloppy commented:
> If you want to fix this, you would need to check in id() if returning real_id is going to result in a change or not. If it wouldn't you would set a bool variable, and then play a sound in loop() if that variable is set.

Ok I have tested this works now, thanks.
I also restored FindBladeAgain() to it's original placement in prop_base.

NoSloppy commented:
> Oh, one more thing; you would want to reset play_no_change_sound_ to false in PlayArraySound().

play_no_change_sound_ will always already be false when reaching PlayArraySound(). It gets set false as soon as it becomes true. Is it ok as I have it? (not set false in PlayArraySound())  Or is this better coding, like a "just incase" defensive thing to add ?

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -1567,6 +1589,25 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+      play_no_change_sound_ = false;
+
+      float real_id = PropBase::id(false);
+      size_t real_best_config = FindBestConfigForId(real_id);
```
if no change happened, then this should be the same as current_config - blade.

NoSloppy replied:
Yes, thanks.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -1567,6 +1589,25 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+      size_t real_best_config = FindBestConfigForId(real_id);
+
+      if (real_best_config == 0) {
+        // Real blade event resulted in no blade array, so blade removal
```
This comment doesn't make any sense.
real_best_config == 0 just means the first entry in the blade array.

profezzorn replied:
Maybe you wanted to check current_config->ohm >= NO_BLADE ?

NoSloppy replied:
This is assumed that NO_BLADE is first in the BladeConfig.
If the blade is removed, that's where we land, so a blade must have been removed.
See comment below.

NoSloppy replied:
> Maybe you wanted to check current_config->ohm >= NO_BLADE ?

This is good.
Done

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -1567,6 +1589,25 @@ class SaberBCButtons : public PROP_INHERIT_PREFIX PropBase {
+
+      if (real_best_config == 0) {
+        // Real blade event resulted in no blade array, so blade removal
+        if (!hybrid_font.PlayPolyphonic(&SFX_bladeout)) {
```
I don't understand why we would play bladeout/bladein if there was no change?
Shouldn't this play the array sound?

NoSloppy replied:
Even if the blade array doesn’t actually change, I'm trying to distinguish between blade insertion and removal.
Yes, this is assumed that NO_BLADE is first in the BladeConfig.
That's how I always run it. I could add a "rule" to the prop comments saying that. But if it's unacceptable then I need to work on it, possibly again unguarding blade_detected_.

profezzorn replied:
So, are you trying to capture the difference between the current state and the previous "fake" state?
Or are you thinking something else?

NoSloppy replied:
Again, I was going with "if a real blade is removed, we will be arriving on NO_BLADE, so must be bladein.wav.
If a blade is inserted, we will be switching to something *not* NO_BLADE, so play bladein.wav."
It is working.
I did change 
```cpp
if (real_best_config == 0) {
        // Real blade event resulted in no blade array, so blade removal
        ```
        
   to use `current_config->ohm >= NO_BLADE` instead, so as not to force NO_BLADE to be the first array.

profezzorn replied:
I think I'm a bit slow today, it makes sense now.

NoSloppy replied:
Oh, well I just found that I still get font.wav if going from an array that's not zero, to another one (like be on array 1 , insert blade, go to array 3, but font.wav plays instead of expected bladein.wav. 
I can't nip it.

profezzorn commented:
> play_no_change_sound_ will always already be false when reaching PlayArraySound(). It gets set false as soon as it becomes true. Is it ok as I have it? (not set false in PlayArraySound()) Or is this better coding, like a "just incase" defensive thing to add ?

It's fine, I was thinking about an edge case, but now that I look at the code, that edge case is nearly impossible to trigger, so it's fine.

profezzorn commented on props/saber_BC_buttons.h:
```diff
@@ -2007,6 +2045,55 @@ void DoSavedTwist() {
+  }
+
+  void NextBladeArray() {
+    best_config_before_faking_ = FindBestConfigForId(PropBase::id(false));
```
I realized that this doesn't work if the blade is on.
Your key bindings are all for MODE_OFF, so maybe it doesn't matter, but the blade can still be on if people have idle effects, long fadeouts, or battery effects.

I think we can replace it with:
```
 if (!use_fake_id_) best_config_before_faking_ = current_config - blades;
```

NoSloppy replied:
Great. In my limited testing, this works the same, so hopefully it helps other scenarios as well.
Edited.
